nutils/numeric.py

```diff
@@ -676,4 +676,20 @@ def sanitize_einsum_subscripts(subscripts, *shapes):
     return (*in_, out)
 
 
+def compress_indices(indices, length):
+    '''Compress an array of monotonically increasing indices.
+
+    The compressed array `c` is such that indices[c[i]:c[i+1]] == i for any
+    value of i in range(length).'''
+
+    step = numpy.empty(len(indices)+1, dtype=int)
```

Comment:
Suggestion: You can save a store to (line 688) and load from `step` (line 689) by using `numpy.subtract(indices[1:], indices[:-1], out=step[1:-1])` as follows:

```python
def compress_indices(indices, length):
    if not len(indices):
         return numpy.zeros((length+1,), int)
     step = numpy.empty(len(indices)+1, dtype=int)
     step[-1] = length - indices[-1]
     numpy.subtract(indices[1:], indices[:-1], out=step[1:-1])
     step[0] = indices[0] + 1
     nz, = step.nonzero()
     return numpy.repeat(nz, step[nz]) # NOTE fails if any step is negative
```

Reply:
Modified, plus added out of bounds check.